contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,276 @@
+  - by level (request headers, request object, storage object)
+
+while
+
+- not degrading apiserver performance when auditing is disabled.
+
+## Constraints and Assumptions
+
+* it is not the goal to implement all output formats one can imagine. The  main goal is to be extensible with a clear golang interface. Implementations of e.g. CADF must be possible, but won't be discussed here.
+* dynamic loading of backends for new output formats are out of scope.
+
+## Use Cases
+
+1. As a cluster operator I want to enable audit logging of requests to the apiserver in order **to comply with given business regulations** regarding a subset of the 7 Ws of auditing:
+
+  - **what** happened?
+  - **when** did it happen?
+  - **who** initiated it?
+  - **on what** did it happen (e.g. pod foo/bar)?
+  - **where** was it observed (e.g. apiserver hostname)?
+  - from **where** was it initiated? (e.g. kubectl IP)
+  - to **where** was it going? (e.g. node 1.2.3.4 for kubectl proxy).
+
+1. Depending on the environment, as a cluster operator I want to **define the amount of audit logging**, balancing computational overhead for the apiserver with the detail and completeness of the log.
+
+1. As a cluster operator I want to **integrate with external systems**, which will have different requirements for the log format, network protocols and communication modes (e.g. pull vs. push).
+
+1. As a cluster operator I must be able to provide a **complete trace of changes to an object** to API objects.
+
+1. As a cluster operator I must be able to create a trace for **all accesses to a secret**.
+
+1. As a cluster operator I want to **define log-rotation** of a file-based output backend.
+
+1. As a cluster operator I must be able to log non-CRUD access like **kubectl exec**, when it started, when it finished and with which initial parameters.
+
+### Out of scope use-cases
+
+1. As a cluster operator I must be able to get a trace of interactive commands executed in **kubectl exec**, **attach** or **run**.
+
+## Community Work
+
+- Kubernetes basic audit log PR: https://github.com/kubernetes/kubernetes/pull/27087/
+- OpenStack's implementation of the CADF standard: https://www.dmtf.org/sites/default/files/standards/documents/DSP2038_1.1.0.pdf
+- Cloud Auditing Data Federation standard: https://www.dmtf.org/standards/cadf
+- Ceilometer audit blueprint: https://wiki.openstack.org/wiki/Ceilometer/blueprints/support-standard-audit-formats
+- Talk from IBM: An Introduction to DMTF Cloud Auditing using
+the CADF Event Model and Taxonomies https://wiki.openstack.org/w/images/e/e1/Introduction_to_Cloud_Auditing_using_CADF_Event_Model_and_Taxonomy_2013-10-22.pdf
+
+## Architecture
+
+When implementing audit logging there are basically two options:
+
+1. put a logging proxy in front of the apiserver
+2. integrate audit logging into the apiserver itself
+
+Both approaches have advantages and disadvanteges:
```

Comment:
Yes, I was talking with @nikhiljindal about it, we'll have to create some sort of UUID which will be passed as a header (or alike) and that will allow us to match audit entries from different apiserver (including federation).